lib/elixir/src/elixir_errors.erl

```diff
@@ -57,6 +61,131 @@ emit_diagnostic(Severity, Position, File, Message, Stacktrace) ->
+
+fancy_exception_snippet(File, LineNumber, Column, Description, Snippet) ->
+  #{content := Content, offset := Offset} = Snippet,
+
+  LineDigits = get_line_number_digits(LineNumber),
+  Spacing = n_spaces(LineDigits + 1),
+
+  Formatted = io_lib:format(
+    " ~ts┌─ ~ts~ts\n"
+    " ~ts│\n"
+    " ~p │ ~ts\n"
+    " ~ts│ ~ts\n"
+    " ~ts│\n"
+    " ~ts~ts",
+    [
+     Spacing, prefix(error), file_format({LineNumber, Column}, File),
+     Spacing,
+     LineNumber, Content,
+     Spacing, highlight_below_line(Content, Offset, error),
+     Spacing,
+     Spacing, format_message(Description, LineDigits)
+    ]),
+
+  unicode:characters_to_binary(Formatted).
+
+fancy_exception(File, LineNumber, Column, Message, ShowLine) ->
+  Formatted = case filelib:is_regular(File) andalso ShowLine of 
+             true -> 
+               line_column_diagnostic({LineNumber, Column}, File, Message, error);
+             false -> 
+               Fmt = no_line_diagnostic(LineNumber, File, Message, error),
+               % Left pad so we stay aligned with "** (Exception)" banner
+               ["   ", string:replace(Fmt, "\n", "\n   ")]
+           end,
+
+  unicode:characters_to_binary(Formatted).
+
+
+no_line_diagnostic(Position, File, Message, Severity) -> 
+  io_lib:format(
+    " ┌─ ~ts~ts\n"
+    " ~ts",
+    [
+     prefix(Severity), file_format(Position, File),
+     Message
+    ]
+   ).
+
+line_column_diagnostic(Position, File, Message, Severity) ->
+  {LineNumber, Column} = Position,
+  Line = get_file_line(File, LineNumber),
+  {TrimmedLine, TotalLeading} = trim_file_line(Line),
+  LineDigits = get_line_number_digits(LineNumber),
+  Spacing = n_spaces(LineDigits + 1),
+
+  io_lib:format(
+    " ~ts┌─ ~ts~ts\n"
+    " ~ts│\n"
+    " ~p │ ~ts\n"
+    " ~ts│ ~ts\n"
+    " ~ts│\n"
+    " ~ts~ts",
+    [
+     Spacing, prefix(Severity), file_format(Position, File),
+     Spacing,
+     LineNumber, TrimmedLine, 
+     Spacing, highlight_below_line(TrimmedLine, Column - TotalLeading, Severity),
+     Spacing,
+     Spacing, format_message(Message, LineDigits)
+    ]
+  ).
+
+format_message(Message, NDigits) ->
+  Padding = list_to_binary(["\n  ", n_spaces(NDigits)]),
+  Bin = unicode:characters_to_binary(Message),
+  binary:replace(Bin, <<"\n">>, Padding, [global]).
+
+highlight_below_line(Line, Column, Severity) ->
+  ErrorLength = match_line_error(Line, Column),
+  case Severity of
+    warning ->  highlight([n_spaces(Column), lists:duplicate(ErrorLength, "~")], warning);
+    error -> highlight([n_spaces(Column), lists:duplicate(ErrorLength, "^")], error)
+  end.
+
+match_line_error(Line, Column) ->
+  TermRegex = "[A-Za-z_\.\{\}\(\)&@]*",
+  {ok, Re} = re:compile(TermRegex),
+  Tail = string:slice(Line, Column),
+ {match, [MatchingTerm]} = re:run(Tail, Re, [{capture, all, binary}]),
+  case MatchingTerm of 
+    % If we cannot consume anything, return a single ^ at column position
+    <<>> -> 1;
+    _ -> string:length(MatchingTerm)
```

Comment:
I am not sure I follow. Could we just highlight the whole line? My concern with trying to infer is that it _will_ fail in certain cases, especially if we are talking about user specific warnings.

Reply:
Yeah, I agree that using a regex or even writing a "mini-parser" function is not 100% reliable, so having to choose between highlighting the whole line and just using a single ^ on the offset, I would go with the latter, since the advantage of having column information is that we can highlight *where* errors happened in the line, and we would lose that by highlighting the whole line.

Reply:
Exactly. If we have a column, we use a single ^, otherwise the whole line. We will be more precise when possible in future diagnostics.